### PR TITLE
Product tests improvements

### PR DESCRIPTION
**Pull request by wendigo**
This PR introduces the following product tests improvements:

* adds logging of the container and environment lifecycle events (starting, started, stopping, stopped)
* refactor codebase to use common, const container names
* adds printing of container stats (cpu/mem/network/pids) before it's stopped
* refactors picocli's Commands to use Callable<Integer> and make exit codes explicit
* adds 4 distinct container output handling modes (DISCARD, PRINT, WRITE, PRINT_WRITE)

### CODE REVIEW

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java [hunk 1]
```diff
@@ -173,6 +261,16 @@ public Builder removeContainer(String name)
...
+        {
+            return build(Optional.empty());
+        }
+
+        public Environment build(EnvironmentListener listener)
+        {
+            return build(Optional.of(listener));
+        }
+
+        protected Environment build(Optional<EnvironmentListener> listener)
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java:281]: private?

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+ * limitations under the License.
+ */
+package io.prestosql.tests.product.launcher.env;
+
+import com.github.dockerjava.api.command.InspectContainerResponse;
+import io.airlift.log.Logger;
+
+import java.util.List;
+
+public interface EnvironmentListener
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java:25]: nit: I would drop "on" prefix from the methods.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java [hunk 3]
```diff
@@ -54,56 +56,56 @@ default void onContainerStopped(DockerContainer container, InspectContainerRespo
     {
     }
 
-    static EnvironmentListener compose(List<EnvironmentListener> listeners)
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java]: this can be moved to commit which introduced `compose`

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java [hunk 4]
```diff
@@ -161,4 +163,24 @@ public void onContainerStopped(DockerContainer container, InspectContainerRespon
             }
         };
     }
+
+    static EnvironmentListener logCopyingListener(Optional<Path> logBaseDir)
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java:170]: make `logBaseDir` non-empty and instead do not create listener in code flow when now you create it with `Optional.empty()`

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java [hunk 5]
```diff
@@ -161,4 +163,24 @@ public void onContainerStopped(DockerContainer container, InspectContainerRespon
...
+        return new EnvironmentListener() {
+            @Override
+            public void onContainerStopping(DockerContainer container, InspectContainerResponse response)
+            {
+                container.copyLogsToHostPath(logBaseDir.get());
+            }
+        };
+    }
+
+    static EnvironmentListener noop()
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java:202]: why do we need this one?

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -298,5 +298,5 @@ jobs:
           rm -rf ~/.m2/repository
       - name: Product Tests
         run: |
-          presto-product-tests-launcher/bin/run-launcher suite run --suite ${{ matrix.suite }} --config ${{ matrix.config }} --no-bind
+          presto-product-tests-launcher/bin/run-launcher suite run --suite ${{ matrix.suite }} --config ${{ matrix.config }} --no-bind --logs-dir logs/
```
Comment (losipiuk) [inline .github/workflows/ci.yml:301]: I think it is fine to leave if it works

**Review by losipiuk — COMMENTED** [review]
Mostly LGTM. Thanks

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java (hunk 5 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java:202]:
```if (logBaseDir.isEmpty()) {
            return noop();
 }```

File: .github/workflows/ci.yml (hunk 6 above)
Comment (wendigo) [reply .github/workflows/ci.yml:301]: 👍🏻

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java (hunk 4 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java:170]: I think it will make code less readable

Comment (losipiuk) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java:170]: But more proper imo. You can create some external helper tooling to construct proper listener set. Logging listener should not allow empty log pth

Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java]: Fixed.

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java [hunk 7]
```diff
@@ -155,26 +153,70 @@ private void copyFileToContainer(String containerPath, Runnable copy)
...
-    private static void verifyHostPath(String hostPath)
+    public void clearDependencies()
     {
-        if (!Files.exists(Paths.get(hostPath))) {
-            throw new IllegalArgumentException("Host path does not exist: " + hostPath);
+        dependencies.clear();
+    }
+
+    public void copyLogsToHostPath(Path hostPath)
+    {
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:166]: This should be in different commit

File: .github/workflows/ci.yml (hunk 6 above)
Comment (losipiuk) [reply .github/workflows/ci.yml:301]: Change commit name

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java [hunk 8]
```diff
@@ -280,20 +293,24 @@ public Environment build(EnvironmentListener listener)
...
-            }
-            catch (UnsupportedEncodingException e) {
-                throw new RuntimeException(e);
+            switch (outputMode) {
+                case DISCARD:
+                    log.warn("Containers logs are not printed to stdout");
+                    setContainerOutputConsumer(this::discardContainerLogs);
+                    break;
+
+                case PRINT:
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java:302]: Can we print to stout also in write mode

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java [hunk 9]
```diff
@@ -304,7 +321,57 @@ private Environment build(Optional<EnvironmentListener> listener)
...
+            // Discard log frames
+            return outputFrame -> {};
+        }
+
+        private void setContainerOutputConsumer(Function<DockerContainer, Consumer<OutputFrame>> consumer)
+        {
+            configureContainers(container -> container.withLogConsumer(consumer.apply(container)));
+        }
+
+        public Builder setStartupRetries(int startupRetries)
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java:365]: Separate commit?

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java (hunk 8 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java:302]: I've added PRINT_WRITE mode 👍🏻

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java (hunk 9 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/Environment.java]: Done

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java (hunk 7 above)
Comment (wendigo) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:166]: Done.

**Comment by findepi** [comment]
>refactor codebase to use common, const container names

what does it mean?

> adds 4 distinct container output handling modes (DISCARD, PRINT, WRITE, PRINT_WRITE)

is it expose in CLI, or internal API in Launcher's env setup?

**Comment by wendigo** [comment]
> > refactor codebase to use common, const container names
> 
> what does it mean?

that instead of using `hadoop-master` across the codebase const string is used instead

> 
> > adds 4 distinct container output handling modes (DISCARD, PRINT, WRITE, PRINT_WRITE)
> 
> is it expose in CLI, or internal API in Launcher's env setup?

in cli when running launcher

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java [hunk 10]
```diff
@@ -73,7 +75,7 @@ public void environmentStarted(Environment environment)
             @Override
             public void environmentStopping(Environment environment)
             {
-                Arrays.stream(listeners).forEach(listener -> listener.environmentStarted(environment));
+                Arrays.stream(listeners).forEach(listener -> listener.environmentStopping(environment));
```
Comment (losipiuk) [inline presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/EnvironmentListener.java:83]: unrelated to commit

File: presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java (hunk 7 above)
Comment (losipiuk) [reply presto-product-tests-launcher/src/main/java/io/prestosql/tests/product/launcher/env/DockerContainer.java:166]: Well - I meant whole migration from `exposeLogsInHOstPath` to log copying. Currently it is part of `Add DockerContainer logical name`.

**Review by losipiuk — COMMENTED** [review]
Overly looks good. It could benefit from some improvements over what change is in which commit.

**Comment by findepi** [comment]
> > > adds 4 distinct container output handling modes (DISCARD, PRINT, WRITE, PRINT_WRITE)
> > 
> > 
> > is it expose in CLI, or internal API in Launcher's env setup?
> 
> in cli when running launcher

and what do the modes mean?

**Comment by wendigo** [comment]
> > > > adds 4 distinct container output handling modes (DISCARD, PRINT, WRITE, PRINT_WRITE)
> > > 
> > > 
> > > is it expose in CLI, or internal API in Launcher's env setup?
> > 
> > 
> > in cli when running launcher
> 
> and what do the modes mean?

* `DISCARD` - discard all output from running containers
* `PRINT` (default, as it's today) - print all output from running containers to stdout
* `WRITE` - write all containers output to log files (under --logs-dir)
* `WRITE_PRINT` - combined WRITE and PRINT behaviors

**Comment by losipiuk** [comment]
Ooops my bad. Merged a bit too soon :/ Codewise it good. But still some commit structure improvements could be made. Sorry.

**Comment by wendigo** [comment]
Can I make a follow-up @losipiuk ?

**Comment by losipiuk** [comment]
> Can I make a follow-up @losipiuk ?

Follow-ups are always welcome :)
